Add Basic Concepts section to docs (#78)
slackapi/bolt-python: docs/_basic/listening_responding_shortcuts.md

---
title: Listening and responding to shortcuts
lang: en
slug: shortcuts
order: 8
---

<div class="section-content">

The `shortcut()` method supports both [global shortcuts](https://api.slack.com/interactivity/shortcuts/using#global_shortcuts) and [message shortcuts](https://api.slack.com/interactivity/shortcuts/using#message_shortcuts).

Shortcuts are invokable entry points to apps. Global shortcuts are available from within search in Slack. Message shortcuts are available in the context menus of messages. Your app can use the `shortcut()` method to listen to incoming shortcut events. The method requires a `callback_id` parameter of type `str` or `RegExp`.

Shortcuts must be acknowledged with `ack()` to inform Slack that your app has received the event.

Shortcuts include a `trigger_id` which an app can use to [open a modal](#creating-modals) that confirms the action the user is taking.

> When configuring shortcuts within your app configuration, you'll continue to append `/slack/events` to your request URL.

Reviewer: ```suggestion
When setting up shortcuts within your app configuration, you'll append `/slack/events` to your request URL.
```

Reviewer: If we want to convey that it's for all URLs, we could make it something like "Like other URLs, you'll append `slack/events` to your request URL when setting up shortcuts."

> ⚠️ Note that global shortcuts do **not** include a channel ID. If your app needs access to a channel ID, you may use a [`conversations_select`](https://api.slack.com/reference/block-kit/block-elements#conversation_select) element within a modal. Message shortcuts do include channel ID.

Reviewer: ```suggestion
⚠️ Note that global shortcuts do **not** include a channel ID. If your app needs access to a channel ID, you may use a [`conversations_select`](https://api.slack.com/reference/block-kit/block-elements#conversation_select) element within a modal. Message shortcuts do include a channel ID.
```

</div>

```python

# The open_modal shortcut opens a plain old modal
@app.shortcut("open_modal")

> def open_modal(ack, client):
>     # Acknowledge the shortcut request
>     ack()
>     # Call the views_open method using one of the built-in WebClients
>     client.views_open(
>         trigger_id=payload["trigger_id"],
>         view={

Reviewer 2: ```suggestion
def open_modal(ack, shortcut, client):
    # Acknowledge the shortcut request
    ack()
    # Call the views_open method using one of the built-in WebClients
    client.views_open(
        trigger_id=shortcut["trigger_id"],
        view={
```

                    }
                },
                {
                    "type": "context",
                    "elements": [
                        {
                            "type": "mrkdwn",
                            "text": "Psssst this modal was designed using <https://api.slack.com/tools/block-kit-builder|*Block Kit Builder*>"
                        }
                    ]
                }
            ]
        }
    )
```

<details class="secondary-wrapper">
  <summary class="section-head" markdown="0">
  <h4 class="section-head">Listening to shortcuts using a constraint object</h4>
  </summary>

  <div class="secondary-content" markdown="0">
  You can use a constraints object to listen to `callback_id`s, and `type`s. Constraints in the object can be of type `str` or `RegExp` object.
  </div>

```python

# Your middleware will only be called when the callback_id matches 'open_modal' AND the type matches 'message_action'

> @app.shortcut({"callback_id": "open_modal", "type": "message_action"})

Reviewer 2: In Bolt for Python, we can go with `@app.message_shortcut("open_modal")`. How about using this? No developer would like to know the type name: message_action here.
```suggestion
@app.message_shortcut("open_modal")
```

> def open_modal(ack, client):
>     # acknowledge the shortcut request
>     ack()
>     # call the views_open method using one of the built-in WebClients
>     client.views_open(
>         trigger_id=payload["trigger_id"],

Reviewer 2: ```suggestion
def open_modal(ack, shortcut, client):
    # acknowledge the shortcut request
    ack()
    # call the views_open method using one of the built-in WebClients
    client.views_open(
        trigger_id=shortcut["trigger_id"],
```
